larq/quantized_variable.py

```diff
@@ -7,7 +7,46 @@
 from larq import quantized_scope
 
 
-class QuantizedVariable(tf.Variable):
+class QuantizedVariableFactory(tf.Variable.__class__):
+    """Factory to allow construction of QuantizedVariable to be overridden."""
+
+    def __call__(cls, variable, quantizer=None, precision=None):
+        """Creates a QuantizedVariable that wraps another variable.
+
+        This typically just returns `QuantizedVariable(variable)`. But, if the variable
+        is a DistributedVariable or one of its subclasses, we instead dynamically
+        create a class that subclasses from both QuantizedVariable and
+        variable.__class__. This is so the returned variable will still pass
+        `isinstance(variable, variable.__class__)`, which is required for
+        DistributedVariables and its subclasses to work properly.
+
+        # Arguments
+        variable: A floating-point resource variable to wrap.
+        quantizer: An optional quantizer to transform the floating-point variable to a
+            fake quantized variable.
+        precision: An optional integer defining the precision of the quantized variable.
+            If `None`, `quantizer.precision` is used.
+
+        # Returns
+        A QuantizedVariable that wraps the variable.
+        """
+
+        if not isinstance(variable, DistributedVariable):  # type: ignore
+            return super().__call__(variable, quantizer, precision)
+
+        class QuantizedDistributedVariable(QuantizedVariable, variable.__class__):
```

Comment:
Apparently this only works for TensorFlow 2.1 and will throw a metaclass conflict otherwise 😞

Reply:
So should we just do a factory instead?